### ISSUE TITLE
Fine tuning lora with SD SD script

### ISSUE THREAD

**Issue by nullarch** [bytedance/ComfyUI-HyperLoRA#35]
Hello. I tried with fine tuning hypelora with SD-script. But I can't get good result. Can you share hyperparams of sd-script for tuning?

**Comment by LemonSky1995** [comment]
Hi, here are some recommended practices:
- Crop the face region to 768x768 before fine-tuning
- Try the following settings:
    - pretrained model: RealVisXL
    - batch size: 4
    - clip skip: 2
    - learning rate: 1e-4
    - don't fine-tune the text encoder

**Comment by nullarch** [comment]
@LemonSky1995 thank you for answer! 
1. did you use prior loss? 
2. is this rank and alpha value is current?
        "--network_dim", "4",
        "--network_alpha", "2",
3. I tested with your config before 5 minutes. but I can get seems untrained result.

**Comment by nullarch** [comment]
@LemonSky1995 

        "accelerate", "launch",
        "--num_processes", "1",
        "--num_cpu_threads_per_process", "2",
        "--main_process_port", "23333",
        "/root/sd-scripts/sdxl_train_network.py",
        "--pretrained_model_name_or_path", "/root/diffusers-models/revis.safetensors",
        "--dataset_config", "config.toml",
        "--sample_prompts", "sample_prompts.txt",
        "--output_dir", f"/root/flux-lora-models/TEST/{train_args['model_uuid']}",
        "--resolution", "768",
        "--save_every_n_epochs", "1",
        "--train_batch_size", "4",
        "--clip_skip", "2",
        "--max_train_epochs", "2",
        "--sample_every_n_epochs", "99999",
        "--learning_rate", "0.0001",
        "--lr_scheduler", "constant",
        "--network_module", "networks.lora",
        "--network_dim", "4",
        "--network_alpha", "2",
        "--no_half_vae",
        "--network_weights", "/root/services/TEST/actor_loras/TEST2.safetensors",
        "--network_train_unet_only",
        "--max_token_length", "225",
        "--bucket_no_upscale",
        "--noise_offset=0.05",
        "--adaptive_noise_scale=0.005",
        "--multires_noise_iterations", "8",
        "--multires_noise_discount", "0.3",
        "--save_model_as", "safetensors",
        "--mixed_precision", "fp16",
        "--save_precision", "fp16",
        "--optimizer_type", "AdamW8bit",
        "--max_grad_norm", "1.0",
        "--keep_tokens", "1",
        "--min_bucket_reso", "256",
        "--max_bucket_reso", "9999999",
        "--persistent_data_loader_workers",
        "--max_data_loader_n_workers", "2"

this is my full train code

**Comment by LemonSky1995** [comment]
> [@LemonSky1995](https://github.com/LemonSky1995) thank you for answer!
> 
> 1. did you use prior loss?
> 2. is this rank and alpha value is current?
>    "--network_dim", "4",
>    "--network_alpha", "2",
> 3. I tested with your config before 5 minutes. but I can get seems untrained result.

prior loss is not necessary. the rank of generated lora is 8, thus network dim and alpha should be 8 and 4. I guess it fails to load the lora weights due to shape mismatches, and downgrade to fine-tuning from scratch.

**Comment by nullarch** [comment]
@LemonSky1995 

I followed the approach you mentioned, but I'm not seeing any learning progress during training. I suspect there might be an issue with how SD scripts loads the LoRA files. Do I need to modify the LoRA loading section in SD scripts?
I'm a machine learning beginner and have been spending a week trying to figure this out, but I'm still unable to identify the problem. Do I need to modify the SD scripts source code to enable fine-tuning of HyperLoRA's output LoRA?
I'm using HyperLoRA very successfully, but if I could figure out how to fine-tune with SD scripts, it would be perfect for my workflow.
In your README.md, you mentioned:

"HyperLoRA can also provide a good initialization for LoRA training. You can obtain a better ID LoRA with only about 50 steps of finetuning."

Could you please share the exact method you used for this 50-step fine-tuning? What specific approach or modifications did you implement?
I'm trying my best to solve this independently to avoid bothering you, but it's proving quite challenging 😅
Thank you for your time and this amazing work!

**Comment by LemonSky1995** [comment]
@SlZeroth hi, we didn't use sd-scripts for finetuning, instead, a new script we wrote. I will test if sd-scripts works later, and give a feedback within a week.

**Comment by nullarch** [comment]
@LemonSky1995 Thank you. Although the HyperLoRA project hasn't received much promotion, in my experience across all projects I've worked on, it's the most effective existing method for proceeding without fine-tuning. I'm looking forward to the additional 50-step fine-tuning method. Once you share it later, I'll support you with a small amount through GitHub Sponsors. Thank you for your hard work

**Comment by zz12138zz** [comment]
I wrote a finetune script with diffusers. To obtain body shape information, I use whole photo instead of just face region to train finetune the lora.In this case, it takes at least 300 steps of training to achieve a reasonably good effect. How is your result? Does using whole photo to finetune make sense?

**Comment by LemonSky1995** [comment]
@SlZeroth Sorry for the late reply. I have tested the latest sd-scripts to finetune the LoRA generated by HyperLoRA. The major reason you failed is that the format of generated LoRA is not compatible with sd-scripts. You can use the following script to convert it to kohya format:
```python
import argparse
from safetensors.torch import load_file, save_file

hyperlora_keys_to_kohya = {
    'lora_unet_down_blocks_1_attentions_0': 'lora_unet_input_blocks_4_1',
    'lora_unet_down_blocks_1_attentions_1': 'lora_unet_input_blocks_5_1',
    'lora_unet_down_blocks_2_attentions_0': 'lora_unet_input_blocks_7_1',
    'lora_unet_down_blocks_2_attentions_1': 'lora_unet_input_blocks_8_1',
    'lora_unet_mid_block_attentions_0': 'lora_unet_middle_block_1',
    'lora_unet_up_blocks_0_attentions_0': 'lora_unet_output_blocks_0_1',
    'lora_unet_up_blocks_0_attentions_1': 'lora_unet_output_blocks_1_1',
    'lora_unet_up_blocks_0_attentions_2': 'lora_unet_output_blocks_2_1',
    'lora_unet_up_blocks_1_attentions_0': 'lora_unet_output_blocks_3_1',
    'lora_unet_up_blocks_1_attentions_1': 'lora_unet_output_blocks_4_1',
    'lora_unet_up_blocks_1_attentions_2': 'lora_unet_output_blocks_5_1'
}

parser = argparse.ArgumentParser(description='Convert HyperLoRA to Kohya format.')
parser.add_argument('--i', type=str, required=True)
parser.add_argument('--o', type=str, required=True)
args = parser.parse_args()

lora = load_file(args.i)
new_lora = {}
for k, v in lora.items():
    items = k.split('_transformer_')
    items[0] = hyperlora_keys_to_kohya[items[0]]
    new_lora['_transformer_'.join(items)] = v
save_file(new_lora, args.o)
```
I prepare 6 images of face region, generate the average LoRA of them, and run the following script to finetune:
```bash
python3 sdxl_train_network.py \
    --pretrained_model_name_or_path "XXX/RealVisXL_v4_BakedVAE.safetensors" \
    --network_weights "XXX.safetensors" \
    --train_data_dir "XXX/data" \
    --output_dir "XXX/saved_lora" \
    --mixed_precision "fp16" \
    --save_precision "fp16" \
    --save_every_n_steps 10 \
    --save_model_as "safetensors" \
    --caption_extension ".txt" \
    --network_module "networks.lora" \
    --network_dim 8 \
    --network_alpha 4 \
    --min_bucket_reso 768 \
    --max_bucket_reso 768 \
    --max_train_steps 50 \
    --train_batch_size 4 \
    --resolution 768 \
    --seed 666 \
    --clip_skip 2 \
    --learning_rate 0.00002 \
    --gradient_checkpointing \
    --network_train_unet_only \
    --enable_bucket \
    --flip_aug \
    --use_8bit_adam \
    --no_half_vae
```

**Comment by LemonSky1995** [comment]
@zz12138zz The ID LoRA barely contains full-body information, so more steps are required for fitting.
